woocommerce/woocommerce-admin
# Fixing a few issues with tracks events in gateway suggestions

Fixes #7303

As per the original issue, some of the tracks events are not being fired quite right. Here is an audit of the events:

## Initial Audit

**wcadmin_task_view**
Currently firing as expected.

> Is it possible to add a new prop showing all the payment methods options the user is seeing when they visit the payments task?

This _would_ be possible, but this event is currently decoupled from the payment gateways suggestions feature entirely, and the data in terms of visible gateways isn't available when fired. Perhaps we could add another event that only fires on loading the payment task?

**wcadmin_tasklist_payment_setup**
This event wasn't firing at all due to an earlier typo 🤭 . This is fixed, although I want to note that the gateway IDs included in the payload _only_ include those that appeared in the same list (Enabled/Additional), and never includes `woocommerce_payments` since it always appears outside of the lists and in it's own card. We could amend this to include all visible gateways if preferred, but I haven't made this change yet.

**wcadmin_tasklist_payment_install_method**
Firing as expected, but only when plugins are installed. This does _not_ include BACS and COD since those don't require installation. This seems like the correct behavior semantically; do we want to change this?

**wcadmin_payments_task_stepper_view**
This seems to be firing as expected. Only when Setup is required for a gateway, and it's being displayed.

**wcadmin_tasklist_payment_connect_start**
This was being fired whenever a connect form was submitted, although by the description it sounds like it's expected to fire when a _Connect_ button directs the user away from the site (Stripe/etc), so I've moved it to execute in this instance only. Let me know if this is not correct.

**wcadmin_tasklist_payment_connect_method**
This occurs for all methods when they're marked as configured within the payments task, as expected I believe.

## Current event flows
_Note: I'll updated this as changes are made in this PR._

**wcadmin_tasklist_payments_options**
Fires when payment task loads, and includes all visible gateway IDs in payload.

**wcadmin_tasklist_payment_setup**
Fires only when "Setup" or "Finish" setup buttons are clicked, so only for gateways that include Setup in the task. Currently includes all IDs of gateways _that appeared in the same list_ (Additional/Enabled/etc), as well as the ID of the selected gateway.

**wcadmin_payments_task_stepper_view**
This fires whenever the setup Stepper is displayed for a gateway. Includes the ID of the gateway.

**wcadmin_tasklist_payment_install_method**
Only fires when a plugin is installed for a gateway as part of the setup. This includes a list of slugs for all plugins installed.

**wcadmin_tasklist_payment_connect_start**
Currently (as of this PR) is firing when a gateway provides a connection URL. This will result in us displaying a "Connect" button during setup, which will fire this event when clicked (and direct the user to the connection URL). Includes ID of gateway.

**wcadmin_tasklist_payment_connect_method**
This is fired whenever a gateway is marked as configured, which _should_ occur for all gateways once they're connected/enabled. Includes ID of gateway.

### Detailed test instructions:

-   Checkout branch
-   Enter this in the console -> `localStorage.setItem( 'debug', 'wc-admin:*' );`
- Go through the basic flows described above for each event and ensure it's fired as expected.
